src/tools/kdb/external.c

```diff
@@ -0,0 +1,205 @@
+#include <sys/stat.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <sys/wait.h>
+#endif
+
+extern char ** environ;
+
+static void insert (StringMap * map, char * key, char * value)
+{
+	for (int i = 0; i < map->size; i++)
+	{
+		if (elektraStrCmp (map->array[i].commandName, key) == 0)
+		{
+			map->array[i].binary = value;
+			return;
+		}
+	}
+	KeyValue newPair;
+	newPair.commandName = key;
+	newPair.binary = value;
+	map->array[map->size++] = newPair;
+}
+
+char * getExternalBin (StringMap * map, const char * key)
+{
+	for (int i = 0; i < map->size; i++)
+	{
+		if (elektraStrCmp (map->array[i].commandName, key) == 0)
+		{
+			return map->array[i].binary;
+		}
+	}
+	return NULL;
+}
+
+void freeExternalBinaryMap (StringMap * map)
+{
+	for (int i = 0; i < map->size; i++)
+	{
+		if (map->array[i].commandName != NULL)
+		{
+			elektraFree (map->array[i].commandName);
+			map->array[i].commandName = NULL;
+		}
+		if (map->array[i].binary != NULL)
+		{
+			elektraFree (map->array[i].binary);
+			map->array[i].binary = NULL;
+		}
+	}
+	map->size = 0;
+}
+
+int tryLoadExternal (char * commandName, StringMap * map)
+{
+	char * execPathPtr = getenv ("KDB_EXEC_PATH");
+	bool found = false;
+	char path[1024] = { 0 };
+	struct stat buf;
+
+	if (execPathPtr)
+	{
+		char * execPath = strdup (execPathPtr);
+		char * token = strtok (execPath, ":");
+		while (token != NULL && !found)
+		{
+			snprintf (path, sizeof (path), "%s/%s", token, commandName);
+			found = stat (path, &buf) != -1;
+			token = strtok (NULL, ":");
+		}
+		elektraFree (execPath);
+	}
+
+	if (!found)
+	{
+		snprintf (path, sizeof (path), "%s/%s", BUILTIN_EXEC_FOLDER, commandName);
+		found = stat (path, &buf) != -1;
+	}
+
+	if (found)
+	{
+		insert (map, elektraStrDup (commandName), strdup (path));
+		return 0;
+	}
+	return 1;
+}
+
+int loadExternalSpec (KeySet * spec, StringMap * map, Key * errorKey)
+{
+	KDB * handle = kdbOpen (NULL, errorKey);
+	Key * baseKey = keyNew ("spec:" CLI_BASE_KEY, KEY_END);
+	KeySet * config = ksNew (0, KS_END);
+	if (kdbGet (handle, config, errorKey) == -1)
+	{
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (errorKey, "could not load '%s': %s", CLI_BASE_KEY, GET_ERR (baseKey));
+		keyDel (baseKey);
+		ksDel (config);
+		kdbClose (handle, errorKey);
+		return 1;
+	}
+	Key * cur = NULL;
+	KeySet * part = ksCut (config, baseKey);
+
+	for (elektraCursor it = 0; it < ksGetSize (part); ++it)
+	{
+		cur = ksAtCursor (part, it);
+		const Key * externalMeta = keyGetMeta (cur, "external");
+		const Key * externalBinary = keyGetMeta (cur, "bin");
+		bool isExternal = false;
+		if (externalBinary != NULL && externalMeta != NULL && elektraKeyToBoolean (externalMeta, &isExternal) && isExternal)
+		{ // add external spec and save path to binary
+			KeySet * externalCommandSpec = ksCut (part, cur);
+			insert (map, elektraStrDup (keyBaseName (cur)), elektraStrDup (keyString (externalBinary)));
+			ksAppend (spec, externalCommandSpec);
+			ksDel (externalCommandSpec);
```

Comment:
```suggestion
			KeySet * externalCommandSpec = ksCut (part, cur);
			insert (map, elektraStrDup (keyBaseName (cur)), elektraStrDup (keyString (externalBinary)));
			ksAppend (spec, externalCommandSpec);
			ksDel (externalCommandSpec);
			--it;
```

Without the `--it` you'd be skipping a key because of `ksCut` (removes `cur` from `part`), I think.